src/cluster.c

```diff
@@ -852,176 +689,306 @@ static dict *parse_cluster_slots(valkeyClusterContext *cc, valkeyReply *reply) {
+static int move_replica_nodes(dict *replicas, dict *nodes) {
+    if (replicas == NULL)
+        return VALKEY_OK;
 
-            part = sdssplitlen(line_start, len + 1, " ", 1, &count_part);
-            if (part == NULL) {
-                goto oom;
+    dictIterator di;
+    dictInitIterator(&di, nodes);
+    dictEntry *de;
+    while ((de = dictNext(&di))) {
+        valkeyClusterNode *primary = dictGetEntryVal(de);
+
+        /* Move all replica nodes related to this primary. */
+        dictEntry *der = dictFind(replicas, primary->name);
+        if (der != NULL) {
+            if (primary->slaves == NULL) {
+                primary->slaves = listCreate();
+                if (primary->slaves == NULL) {
+                    return VALKEY_ERR;
+                }
+                primary->slaves->free = listClusterNodeDestructor;
             }
 
-            if (count_part < 8) {
-                valkeyClusterSetError(cc, VALKEY_ERR_OTHER,
-                                      "split cluster nodes error");
-                goto error;
+            /* Move all replicas stored in the first parsed replica. */
+            valkeyClusterNode *replica = dictGetEntryVal(der);
+            if (replica->slaves != NULL) {
+                while (listLength(replica->slaves) > 0) {
+                    listNode *node = listFirst(replica->slaves);
+                    if (listAddNodeTail(primary->slaves, node->value) == NULL) {
+                        return VALKEY_ERR;
+                    }
+                    /* Delete element without freeing the moved cluster node. */
+                    replica->slaves->free = NULL;
+                    listDelNode(replica->slaves, node);
+                    replica->slaves->free = listClusterNodeDestructor;
+                }
+                listRelease(replica->slaves);
+                replica->slaves = NULL;
+            }
+            /* Move replica that was parsed first. */
+            if (listAddNodeHead(primary->slaves, replica) == NULL) {
+                return VALKEY_ERR;
             }
+            /* All replicas for this primary moved, set dict value
+             * to NULL avoiding freeing the moved memory. */
+            dictSetHashVal(replicas, der, NULL);
+        }
+    }
+    return VALKEY_OK;
+}
 
-            // if the address string starts with ":0", skip this node.
-            if (sdslen(part[1]) >= 2 && memcmp(part[1], ":0", 2) == 0) {
-                sdsfreesplitres(part, count_part);
-                count_part = 0;
-                part = NULL;
+/* Parse a node from a single CLUSTER NODES line. Only parse primary nodes if
+ * the 'replica_master_id' argument is NULL, otherwise replicas are parsed and
+ * its master_id is given via 'replica_master_id'. */
+static int parse_cluster_nodes_line(valkeyClusterContext *cc, char *line,
+                                    valkeyClusterNode **parsed_node, char **replica_master_id) {
```

Comment:
"Given" usually means input parameter, as in given by the caller, but this is returned by pointer, right? Or maybe I don't really understand what you mean here.

If it is return-by-pointer, I usually document is as something like "replica_master_id is pointed to the master id", or just "returned by pointer".

Please document `parsed_node` too. (It is pointed to the parsed node, right?)